# 1. Import Libraries and Packages

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# 2. Data Preparation

In [3]:
data = pd.read_csv('concrete_data.csv')
data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


In [5]:
data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

# A. Build a Baseline Model

In [9]:
predictors = data.drop('Strength',axis=1)
target = data['Strength']

In [10]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(predictors, target, test_size = 0.3, random_state = 1)

In [11]:
import keras

In [12]:
from keras.models import Sequential
from keras.layers import Dense

In [16]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [17]:
n_cols = predictors.shape[1] # number of predictors
mean_squared_errors = []

for i in range(50):
    x_train, x_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=1)
    model = regression_model()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=50, verbose=2)
    mse = model.evaluate(x_test, y_test, verbose=0)
    mean_squared_errors.append(mse)

# Compute the mean and standard deviation of the mean squared errors
import numpy as np
mean_mse = np.mean(mean_squared_errors)
std_mse = np.std(mean_squared_errors)

print("Mean of Mean Squared Errors: {}".format(mean_mse))
print("Standard Deviation of Mean Squared Errors: {}".format(std_mse))


Streaming output truncated to the last 5000 lines.
Epoch 2/50
23/23 - 0s - loss: 31881.7949 - val_loss: 10716.9326 - 104ms/epoch - 5ms/step
Epoch 3/50
23/23 - 0s - loss: 6511.1655 - val_loss: 4212.2378 - 112ms/epoch - 5ms/step
Epoch 4/50
23/23 - 0s - loss: 4462.5195 - val_loss: 4045.1658 - 162ms/epoch - 7ms/step
Epoch 5/50
23/23 - 0s - loss: 4096.5938 - val_loss: 3650.3804 - 195ms/epoch - 8ms/step
Epoch 6/50
23/23 - 0s - loss: 3835.0630 - val_loss: 3376.7698 - 217ms/epoch - 9ms/step
Epoch 7/50
23/23 - 0s - loss: 3573.9011 - val_loss: 3131.2253 - 265ms/epoch - 12ms/step
Epoch 8/50
23/23 - 0s - loss: 3335.2209 - val_loss: 2882.7517 - 228ms/epoch - 10ms/step
Epoch 9/50
23/23 - 0s - loss: 3106.2898 - val_loss: 2673.7202 - 171ms/epoch - 7ms/step
Epoch 10/50
23/23 - 0s - loss: 2889.7473 - val_loss: 2442.0210 - 131ms/epoch - 6ms/step
Epoch 11/50
23/23 - 0s - loss: 2676.0542 - val_loss: 2252.3267 - 223ms/epoch - 10ms/step
Epoch 12/50
23/23 - 0s - loss: 2480.9731 - val_loss: 2077.5620 - 225ms/e

In [21]:
# Compute the mean and standard deviation of the mean squared errors
mean_mse = np.mean(mean_squared_errors)
std_mse = np.std(mean_squared_errors)

print("Mean of Mean Squared Errors: {}".format(mean_mse))
print("Standard Deviation of Mean Squared Errors: {}".format(std_mse))

Mean of Mean Squared Errors: 379.1533671569824
Standard Deviation of Mean Squared Errors: 361.7602202528212


Comment:
- The mean of the Mean Squared Errors for the Baseline model is 379.1533671569824 .
- The standard deviation of the Mean Squared Errors for the Baseline model is 361.7602202528212 .

# B. Normalize the Data

In [22]:
predictors = data.drop('Strength',axis=1)
target = data['Strength']

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [23]:
n_cols = predictors_norm.shape[1] # number of predictors
mean_squared_errors = []

for i in range(50):
    x_train, x_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.3, random_state = 1)
    model = regression_model()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=50, verbose=2)
    mse = model.evaluate(x_test, y_test, verbose=0)
    mean_squared_errors.append(mse)

Epoch 1/50
23/23 - 1s - loss: 1494.9760 - val_loss: 1533.2015 - 778ms/epoch - 34ms/step
Epoch 2/50
23/23 - 0s - loss: 1477.8743 - val_loss: 1516.3651 - 92ms/epoch - 4ms/step
Epoch 3/50
23/23 - 0s - loss: 1460.1716 - val_loss: 1498.5330 - 86ms/epoch - 4ms/step
Epoch 4/50
23/23 - 0s - loss: 1441.4147 - val_loss: 1479.8051 - 91ms/epoch - 4ms/step
Epoch 5/50
23/23 - 0s - loss: 1421.6385 - val_loss: 1460.0364 - 93ms/epoch - 4ms/step
Epoch 6/50
23/23 - 0s - loss: 1400.5651 - val_loss: 1438.6678 - 99ms/epoch - 4ms/step
Epoch 7/50
23/23 - 0s - loss: 1377.8544 - val_loss: 1416.7031 - 94ms/epoch - 4ms/step
Epoch 8/50
23/23 - 0s - loss: 1354.4619 - val_loss: 1392.4821 - 85ms/epoch - 4ms/step
Epoch 9/50
23/23 - 0s - loss: 1329.3529 - val_loss: 1367.5320 - 96ms/epoch - 4ms/step
Epoch 10/50
23/23 - 0s - loss: 1303.2010 - val_loss: 1341.2361 - 89ms/epoch - 4ms/step
Epoch 11/50
23/23 - 0s - loss: 1275.5564 - val_loss: 1313.9440 - 156ms/epoch - 7ms/step
Epoch 12/50
23/23 - 0s - loss: 1246.7427 - val_lo

In [24]:
# Compute the mean and standard deviation of the mean squared errors
mean_mse = np.mean(mean_squared_errors)
std_mse = np.std(mean_squared_errors)

print("Mean of Mean Squared Errors: {}".format(mean_mse))
print("Standard Deviation of Mean Squared Errors: {}".format(std_mse))

Mean of Mean Squared Errors: 392.5326770019531
Standard Deviation of Mean Squared Errors: 111.44369752562872


Comment:

- The mean of the Mean Squared Errors for the Normalized model is 392.5326770019531 .
- The standard deviation of the Mean Squared Errors for the Normalized model is 111.44369752562872 .
- When comparing the normalized model to Part A, it appears that the Mean Squared Errors have slightly increased, while the standard deviation has decreased.

# C. Increase the number of Epochs

- Use epochs = 100

In [25]:
predictors = data.drop('Strength',axis=1)
target = data['Strength']

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [26]:
n_cols = predictors_norm.shape[1] # number of predictors
mean_squared_errors = []

for i in range(50):
    x_train, x_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.3, random_state = 1)
    model = regression_model()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=100, verbose=2)
    mse = model.evaluate(x_test, y_test, verbose=0)
    mean_squared_errors.append(mse)

Streaming output truncated to the last 5000 lines.
Epoch 1/100
23/23 - 1s - loss: 1487.4781 - val_loss: 1512.8645 - 855ms/epoch - 37ms/step
Epoch 2/100
23/23 - 0s - loss: 1468.1229 - val_loss: 1492.3749 - 107ms/epoch - 5ms/step
Epoch 3/100
23/23 - 0s - loss: 1448.1000 - val_loss: 1471.2894 - 121ms/epoch - 5ms/step
Epoch 4/100
23/23 - 0s - loss: 1428.0215 - val_loss: 1448.7698 - 145ms/epoch - 6ms/step
Epoch 5/100
23/23 - 0s - loss: 1406.8739 - val_loss: 1426.3955 - 110ms/epoch - 5ms/step
Epoch 6/100
23/23 - 0s - loss: 1385.1477 - val_loss: 1402.9973 - 117ms/epoch - 5ms/step
Epoch 7/100
23/23 - 0s - loss: 1362.5376 - val_loss: 1379.0557 - 116ms/epoch - 5ms/step
Epoch 8/100
23/23 - 0s - loss: 1339.0806 - val_loss: 1353.8885 - 191ms/epoch - 8ms/step
Epoch 9/100
23/23 - 0s - loss: 1314.8395 - val_loss: 1327.6276 - 167ms/epoch - 7ms/step
Epoch 10/100
23/23 - 0s - loss: 1289.5599 - val_loss: 1300.7131 - 153ms/epoch - 7ms/step
Epoch 11/100
23/23 - 0s - loss: 1263.5698 - val_loss: 1273.0450 - 1

In [27]:
# Compute the mean and standard deviation of the mean squared errors
mean_mse = np.mean(mean_squared_errors)
std_mse = np.std(mean_squared_errors)

print("Mean of Mean Squared Errors: {}".format(mean_mse))
print("Standard Deviation of Mean Squared Errors: {}".format(std_mse))

Mean of Mean Squared Errors: 192.7166961669922
Standard Deviation of Mean Squared Errors: 24.697531947985365


Comment:

- The mean of the Mean Squared Errors for the Normalized model is 192.7166961669922 .
- The standard deviation of the Mean Squared Errors for the Normalized model is 24.697531947985365 .
- When comparing the normalized model to Part B, it appears that the Mean Squared Errors have decreased, while the standard deviation has also decreased when the epoch is increased.

# D. Increase the number of hidden layers

Repeat part B (set epoch=50) but with updated regression model

In [31]:
predictors = data.drop('Strength',axis=1)
target = data['Strength']

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [32]:
# define regression model with 3 hidden layers with 10 nodes each
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [33]:
n_cols = predictors_norm.shape[1] # number of predictors
mean_squared_errors = []

for i in range(50):
    x_train, x_test, y_train, y_test = train_test_split(predictors_norm, target, test_size = 0.3, random_state = 1)
    model = regression_model()
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=50, verbose=2)
    mse = model.evaluate(x_test, y_test, verbose=0)
    mean_squared_errors.append(mse)

Epoch 1/50
23/23 - 1s - loss: 1552.7880 - val_loss: 1578.3273 - 1s/epoch - 48ms/step
Epoch 2/50
23/23 - 0s - loss: 1525.6272 - val_loss: 1547.4855 - 102ms/epoch - 4ms/step
Epoch 3/50
23/23 - 0s - loss: 1495.7119 - val_loss: 1510.8540 - 115ms/epoch - 5ms/step
Epoch 4/50
23/23 - 0s - loss: 1459.3163 - val_loss: 1467.1587 - 100ms/epoch - 4ms/step
Epoch 5/50
23/23 - 0s - loss: 1415.3351 - val_loss: 1413.1288 - 102ms/epoch - 4ms/step
Epoch 6/50
23/23 - 0s - loss: 1361.0870 - val_loss: 1346.0664 - 109ms/epoch - 5ms/step
Epoch 7/50
23/23 - 0s - loss: 1292.9220 - val_loss: 1264.7898 - 119ms/epoch - 5ms/step
Epoch 8/50
23/23 - 0s - loss: 1211.7371 - val_loss: 1166.3021 - 104ms/epoch - 5ms/step
Epoch 9/50
23/23 - 0s - loss: 1112.5042 - val_loss: 1055.4913 - 91ms/epoch - 4ms/step
Epoch 10/50
23/23 - 0s - loss: 1001.2832 - val_loss: 925.3632 - 97ms/epoch - 4ms/step
Epoch 11/50
23/23 - 0s - loss: 874.7086 - val_loss: 791.0883 - 97ms/epoch - 4ms/step
Epoch 12/50
23/23 - 0s - loss: 743.8832 - val_los

In [34]:
# Compute the mean and standard deviation of the mean squared errors
mean_mse = np.mean(mean_squared_errors)
std_mse = np.std(mean_squared_errors)

print("Mean of Mean Squared Errors: {}".format(mean_mse))
print("Standard Deviation of Mean Squared Errors: {}".format(std_mse))

Mean of Mean Squared Errors: 172.97255157470704
Standard Deviation of Mean Squared Errors: 8.19303996546273


Comment:

- The mean of the Mean Squared Errors for the Normalized model is 172.97255157470704 .
- The standard deviation of the Mean Squared Errors for the Normalized model is 8.19303996546273 .
- When comparing the normalized model to Part B, it appears that the Mean Squared Errors have decreased, while the standard deviation has also decreased when more hidden layers are added.